ipa/ipa_ec2.py

```diff
@@ -259,34 +231,92 @@ def _get_user_data(self):
+        """
+        Return True if instance is in running state.
+        """
+        return self._get_instance_state() == 'running'
+
     def _launch_instance(self):
         """Launch an instance of the given image."""
+        resource = self._connect()
+
+        instance_name = ipa_utils.generate_instance_name('ec2-ipa-test')
         kwargs = {
-            'name': ipa_utils.generate_instance_name('ec2-ipa-test'),
-            'size': self._get_instance_size(),
-            'image': self._get_image()
+            'InstanceType': self.instance_type or EC2_DEFAULT_TYPE,
+            'ImageId': self.image_id,
+            'MaxCount': 1,
+            'MinCount': 1,
+            'TagSpecifications': [
+                {
+                    'ResourceType': 'instance',
+                    'Tags': [
+                        {
+                            'Key': 'Name',
+                            'Value': instance_name
+                        }
+                    ]
+                }
+            ]
         }
 
         if self.ssh_key_name:
-            kwargs['ex_keyname'] = self.ssh_key_name
+            kwargs['KeyName'] = self.ssh_key_name
         else:
-            kwargs['ex_userdata'] = self._get_user_data()
+            kwargs['UserData'] = self._get_user_data()
 
         if self.subnet_id:
-            kwargs['ex_subnet'] = self._get_subnet(self.subnet_id)
+            kwargs['SubnetId'] = self.subnet_id
 
         if self.security_group_id:
-            kwargs['ex_security_group_ids'] = [self.security_group_id]
+            kwargs['SecurityGroupIds'] = [self.security_group_id]
 
-        instance = self.compute_driver.create_node(**kwargs)
+        try:
+            instances = resource.create_instances(**kwargs)
+        except Exception as error:
+            raise EC2ProviderException(
+                'Unable to create instance: {0}.'.format(error)
+            )
 
-        self.compute_driver.wait_until_running(
-            [instance],
-            timeout=self.timeout
-        )
-        self.running_instance_id = instance.id
+        instances[0].wait_until_running()
+        self.running_instance_id = instances[0].instance_id
 
     def _set_image_id(self):
         """If existing image used get image id."""
         instance = self._get_instance()
-        self.image_id = instance.extra['image_id']
+        self.image_id = instance.image_id
+
+    def _set_instance_ip(self):
+        """
+        Get instance ip of instance.
+
+        Current prefrence is for ipv4.
+        """
+        instance = self._get_instance()
+        if instance.public_ip_address:
+            self.instance_ip = instance.public_ip_address
+        else:
+            try:
+                self.instance_ip = \
+                    instance.network_interfaces[0].ipv6_addresses[0]
+            except IndexError:
+                raise EC2ProviderException(
+                    'IP address for instance cannot be found.'
+                )
```

Comment:
This implementation is not compatible, i.e. will trigger an error, with an instance that runs in a VPC that has IPv4 only and uses only a private IP address. We need to work a little harder.

Reply:
I think if an instance has a network interface it should have at least a private ip. Which is accessible via instance.private_ip_address. The default public address (ipv4 only) should be instance.public_ip_address. And if there's only an ipv6 address the only way to retrieve it is from an interface list. interface.ipv6_addresses. Will add private IP as third option.